Add vignette for odin.dust, dust and mcstate (#76)
mrc-ide/mcstate: vignettes/sir_models.Rmd

The anatomy of an mcstate particle filter, as noted above, consists of three main components:

- A set of observations to fit the model to, generated using `mcstate::particle_filter_data()`.
- A model to fit, which must be a dust generator, either `dust::dust()` or `odin.dust::odin_dust()`.
- A comparison function, which is an R function which calculates the likelihood of the state given the data at one time point.

## Model data

To demonstrate this, we will attempt to infer the parameters $\beta$ and $\gamma$ using daily case counts. In reality these counts would be measured, but here we have generated them from a single run of the model with `seed = 2L`, and added some constant noise proportional to the epidemic size:

```{r}
# Set up model

> incidence <- t(read.table("sir_incidence_data.txt", sep=",", header = F))

Reviewer: See the above comment about where this data should come from

incidence <- data.frame(cases = incidence[, 1], day = seq_len(100), row.names = NULL)
```

This can then be processed with `mcstate::particle_filter_data()` as follows, taking four update steps between every observation (generally decreasing $dt$ will improve resolution, especially if rates of change are high, at the expense of extra computation time):

```{r paged.print=TRUE, layout="l-body-outset", fig.height=5, fig.width=7}
dt <- 0.25
sir_data <- mcstate::particle_filter_data(data = incidence,
                                          time = "day",

>                                           rate = 1/dt)

Reviewer: This use of `dt` begs for more explanation

}
```

We can use the `info()` method on a dust model to inspect what order the state variables will come in:

```{r}
gen_sir$new(data = list(), step = 0, n_particles = 1L)$info()
```

The SIR example actually outputs incidence directly as part of the state, so the comparison function can be written without needing to recalculate this:

```{r}
incidence_compare <- function(state, prev_state, observed, pars = NULL) {
  exp_noise <- 1e6

  lambda <- state[4, , drop = TRUE] +
    rexp(n = length(incidence_modelled), rate = exp_noise)
  dpois(x = observed$cases, lambda = lambda, log = TRUE)
}
```

Just some of the state can be extracted for the comparison function, which is demonstrated below.

## Inferring parameters

Using these pieces, we can set up a particle filter as follows:

```{r}
n_particles <- 100
filter <- mcstate::particle_filter$new(data = sir_data,
                                       model = gen_sir,
                                       n_particles = n_particles,
                                       compare = case_compare,
                                       n_threads = 4L,
                                       seed = 1L)
```

> We can now run the particle filter forward, which will run and resample 100 trajectories, and return the final likelihood. It is important to set the correct time-step $dt$ here, as we are using 0.25, rather than the default value of 1 defined in the odin model. We will also save the history, which allows us to plot the particle trajectories, as well as use the likelihood:

Reviewer: `$dt` should become `dt` in backticks for consistency

```{r}
proposal_matrix <- cov(pmcmc_run$pars)
mcmc_pars <- mcstate::pmcmc_parameters$new(list(beta = beta, gamma = gamma), proposal_matrix)
proposal_matrix
```

Let's now run four independent chains with these proposals:

```{r, fig.height=10, fig.width=14}
pmcmc_tuned_run <-
  mcstate::pmcmc(
    mcmc_pars,
    filter,
    2000,
    save_state = TRUE,
    save_trajectories = TRUE,
    progress = TRUE,
    n_chains = 4
  )

mcmc2 <- coda::as.mcmc(cbind(pmcmc_tuned_run$probabilities, pmcmc_tuned_run$pars))

summary(mcmc2)
plot(mcmc2)
```

Looking again at the effective sample size and rejection rate, these chains were a lot more efficient, and have a much better effective sample size thanks for the four chains:

```{r}
effectiveSize(mcmc2)
1 - rejectionRate(mcmc2)
```

## Running predictions

Going back to the simple SIR model above, we can easily continue running the fitted model forward in time using the posterior estimated by `pmcmc()` to create a forecast past the end of the data:

```{r fig.height=5, fig.width=7}
forecast <- predict(processed_chains,
                    steps = seq(400, 800, 4),

>                     n_threads = 4L,

Reviewer: throughout this vignette, please remove all uses of the `n_threads` argument.

The reasons for doing this are that we need to not use more than 2 threads on CRAN, but also it complicates writing the parallel vignette

                    seed = processed_chains$predict$seed)

# Sample 10 particles
mini_forecast <- forecast$state[,sample.int(ncol(forecast$state), size = 10),]

plot_particle_filter(mini_forecast,
                     true_history,
                     seq_len(200),
                     obs_end = max(incidence$day))
```

Up to $t = 100$ is the nowcast, from $t = 100$ to $t = 200$ is the forecast.

## Fitting to multiple datastreams

The set up of the particle filter and observation function makes it easy to fit to multiple datastreams within a Bayesian framework. Let us model a disease where a proportion of infected cases $p_{\mathrm{death}} = 0.05$ die rather than recover by adding the following lines to the odin code:

```{r pkg_dust_r, echo = FALSE, results = "asis"}
r_output(readLines(file.path("deaths.R")))
```

```{r}
gen_death <- odin.dust::odin_dust("deaths.R")
```

We can simulate some data to fit to, as before. We have also made the case data patchy, so it is missing on some days:

```{r layout="l-body-outset", paged.print=TRUE }
# Set up model
dt <- 0.25
death_data <- read.table("death_data.csv", header = TRUE, sep = ",")

> true_history_deaths <- readRDS("deaths_true_history.rds")

Reviewer: All these rds files need saving in the same script as the .csv - can you save them with `saveRDS(..., version = 2)` so that we use a version of rds that works with older R?

---
title: "SIR models with odin, dust and mcstate"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{sir_models}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

> ```{r setup}

Reviewer: I know opinions vary, but I'm in favour of not explicitly loading the libraries

library(odin.dust)
library(mcstate)
library(rmarkdown)
```

> 

Reviewer: Probably we need some basic introductory stuff here about what the problem is we're trying to solve

# Stochastic SIR model definition

A simple definition of the SIR model, as given in the [odin documentation](https://mrc-ide.github.io/odin/articles/discrete.html) is:
$$\begin{align*} 
\frac{dS}{dt} &= -\beta \frac{SI}{N} \\
\frac{dI}{dt} &= \beta \frac{SI}{N} - \gamma I \\
\frac{dR}{dt} &= \gamma I \\
\end{align*}$$
$S$ is the number of susceptibles, $I$ is the number of infected and $R$ is the number recovered; the total population size $N = S + I + R$ is constant. $\beta$ is the infection rate, $\gamma$ is the recovery rate.

Discretising this model in time steps of width $dt$ gives the following update equations for each time step:

$$\begin{align*} 
S_{t+1} &= S_t - n_{SI} \\
I_{t+1} &= I_t + n_{SI} - n_{IR} \\
R_{t+1} &= R_t - n_{IR}
\end{align*}$$

where
$$\begin{align*} 
n_{SI} &\sim B(S, 1 - e^{-\beta \frac{I}{N} \cdot dt)} \\
n_{IR} &\sim B(I, 1 - e^{-\gamma \cdot dt)}
\end{align*}$$

> 

Reviewer: For the *very first* example, I'd wondered about using one of the builtin examples. So we have already `dust::dust_example("sir")` which implements this model. The reasoning here is that it's less stuff to start getting started with

## Implementing the SIR model using [`odin.dust`](https://mrc-ide.github.io/odin.dust/)

The above equations can straightforwardly be written out using the odin DSL:

> ```r

Reviewer: This duplicates your model file, which will lead to confusion. Can you adapt the approach in dust to read in and print to screen.  Look for the part that says

````
```{r pkg_dust_r, echo = FALSE, results = "asis"}
r_output(readLines(file.path(path, "R/dust.R")))
```
````

## Definition of the time-step and output as "time"
dt <- user(1)
initial(time) <- 0
update(time) <- (step + 1) * dt

## Core equations for transitions between compartments:
update(S) <- S - n_SI
update(I) <- I + n_SI - n_IR
update(R) <- R + n_IR

## Individual probabilities of transition:
p_SI <- 1 - exp(-beta * I / N * dt) # S to I
p_IR <- 1 - exp(-gamma * dt) # I to R

## Draws from binomial distributions for numbers changing between
## compartments:
n_SI <- rbinom(S, p_SI)
n_IR <- rbinom(I, p_IR)

## Total population size
N <- S + I + R

## Initial states:
initial(S) <- S_ini
initial(I) <- I_ini
initial(R) <- 0

## User defined parameters - default in parentheses:
S_ini <- user(1000)
I_ini <- user(10)
beta <- user(0.2)
gamma <- user(0.1)
```

This is converted to a C++ dust model, and compiled into a library in a single step, using [`odin.dust`](https://mrc-ide.github.io/odin.dust/). Save the above code as a file named `sir.R`. File names must not contain special characters.

```{r}
library(odin.dust)

> gen_sir <- odin_dust("sir.txt")

Reviewer: I'd prefer an explicit `odin.dust::odin_dust()` here

```

> ## Adding age structure to the model

Reviewer: This feels like it's in the wrong vignette to me

- Where an age compartment needs to be reduced into a single compartment/variable, use `sum` (or another array function from https://mrc-ide.github.io/odin/articles/functions.html as appropriate)
- Define the dimensions of all arrays, for example by setting `dim(S) <- N_age`.

This would simply give `N_age` independent processes equivalent to the first model, scaled by the size of the population in each age category. To actually make this useful, you likely want to add some interaction or transitions between the compartments. An example of this would be to add an age-specific contact matrix, demonstrated below, which defines a different force of infection $\lambda$ for each age group. This is calculated by $$\lambda_i = \frac{\beta}{N} \cdot \sum_{j=1}^{N_{\mathrm{age}}} I_j m_{ij}$$ In the odin code:

```r
m[, ] <- user() # age-structured contact matrix
s_ij[, ] <- m[i, j] * I[i]
lambda[] <- beta / N * sum(s_ij[i, ])
```

The probability of infection of a susceptible is then indexed by this force of infection:

```r
p_SI[] <- 1 - exp(-lambda[i] * dt)
```

Putting this all together, the age structured SIR model is as follows:

> ```r

Reviewer: This is not really used in the above? It looks like it would fit better in the odin vignette to me

dt <- user(1)
initial(time) <- 0
update(time) <- (step + 1) * dt

## Core equations for transitions between compartments:
update(S[]) <- S[i] - n_SI[i]
update(I[]) <- I[i] + n_SI[i] - n_IR[i]
update(R[]) <- R[i] + n_IR[i]

## Individual probabilities of transition:
p_SI[] <- 1 - exp(-lambda[i] * dt) # S to I
p_IR <- 1 - exp(-gamma * dt) # I to R

## Force of infection
m[, ] <- user() # age-structured contact matrix
s_ij[, ] <- m[i, j] * I[i]
lambda[] <- beta / N * sum(s_ij[i, ])

## Draws from binomial distributions for numbers changing between
## compartments:
n_SI[] <- rbinom(S[i], p_SI[i])
n_IR[] <- rbinom(I[i], p_IR)

## Total population size
N <- sum(S) + sum(I) + sum(R)

## Initial states:
initial(S[]) <- S_ini
initial(I[]) <- I_ini
initial(R[]) <- 0

## User defined parameters - default in parentheses:
S_ini <- user(1000)
I_ini <- user()
beta <- user(0.2)
gamma <- user(0.1)

# dimensions of arrays
N_age <- user()
dim(S) <- N_age
dim(I) <- N_age
dim(R) <- N_age
dim(n_SI) <- N_age
dim(n_IR) <- N_age
dim(p_SI) <- N_age
dim(m) <- c(N_age, N_age)
dim(s_ij) <- c(N_age, N_age)
dim(lambda) <- N_age
```

As before, save the file, and use `odin.dust` to compile the model:

```{r}

> gen_age <- odin_dust("sirage.txt")

Reviewer: This appears never to be used? I'd remove this section entirely for now

```

We can generate an age-structured contact matrix based on the [POLYMOD](https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.0050074) survey by using the [socialmixr](https://github.com/sbfnk/socialmixr) package:

```{r}
library(socialmixr)
data(polymod)

age.limits = seq(0, 70, 10)

## Get the contact matrix from socialmixr
contact <- socialmixr::contact_matrix(
  survey = polymod,
  countries = "United Kingdom",
  age.limits = age.limits,
  symmetric = TRUE)

## Transform the matrix to the (symetrical) transmission matrix
## rather than the contact matrix
transmission <- contact$matrix /
  rep(contact$demography$population, each = ncol(contact$matrix))
transmission
```

This can be given as a parameter to the model by adding the argument `data = list(m = transmission)` when using the `new()` method on the `gen_age` generator.

> ## Saving a model into a package

Reviewer: This is definitely in the wrong vignette - this belongs in odin.dust

If you want to distribute your model in an R package, rather than building it locally, you will want to use the `odin_dust_package()` function instead. This will write the transpiled C++ dust code into `src` and its R interface in `R/dust.R`. Package users are not required to regenerate the dust code, and their compiler will build the library when they install the package. You may also wish to add a file `R/zzz.R` to your package `mycoolpackage` with the following lines:

```r
##' @useDynLib mycoolpackage, .registration = TRUE
NULL
```

to help automate the compilation of the model.

> # Running the SIR model with [`dust`](https://mrc-ide.github.io/dust/)

Reviewer: IMO all the material above here can be dropped and/or moved to different vignettes. I think as much as possible this vignette should focus on the "I want to do inference" problem probably. We can have as many vignettes as needed and IME keeping them fairly focussed, while hard, is necessary to keep them clear

Author: Moving everything above the mcstate section to odin.dust

Now we can use the `new` method on the generator to make `dust` objects. Typically you will work through the `mcstate` package to run the model, but for illustrative purposes we show here that `dust` can be driven directly from R.

[`new()`](https://mrc-ide.github.io/dust/reference/dust.html#method-new) takes the data needed to run the model i.e. a list with any parameters defined as `user` in the odin code above, the value of the initial step $t_0$, and the number of particles, each for now can simply be thought of as an independent stochastic realisation of the model, but in the next step will be used when inferring model parameters.

> Additional arguments include the number of threads to parallelise the particles over, and the seed for the random number generator. The seed must be an integer, and using the same seed will ensure reproducible results for all particles. For threads to take effect, `openmp` must be enabled in the compilation step, which you can check by running:

Reviewer: Detail like this probably goes later on in a section on peformance?

```{r}
gen_sir$public_methods$has_openmp()
```

To use this to directly create a new dust object with 10 particles, run using 4 threads:

```{r}
sir_model <- gen_sir$new(data = list(dt = 1, S_ini = 1000, I_ini = 10, beta = 0.2, gamma = 0.1),
                         step = 1,
                         n_particles = 10L,
                         n_threads = 4L,
                         seed = 1L)
```

> Check the initial state is ten particles wide, with four states deep ($t$, $S$, $I$, $R$):

Reviewer: ```suggestion
The initial state is ten particles wide, four states deep ($t$, $S$, $I$, $R$):
```

```{r}
sir_model$state()
```

Run the particles (repeats) forward 10 steps of length $dt$, followed by another 10 steps:

```{r}
sir_model$run(10)
sir_model$run(20)
```

> To change the parameters, the model does not need to be recompiled. We simply use the generator object to make a new model. Let's do this and increase the infection rate, and the population size. We will also use a smaller time step to calculate multiple transitions per unit time:

Reviewer: I think that the compilation part here is a red-herring (not the detail the user will be expecting). You might write this as

```suggestion
We can change the parameters, say by increasing the infection rate and the population size, by reinitalising the model. We will also use a smaller time step to calculate multiple transitions per unit time:
```

**However** this goes against advice in the dust vignette to use `$reset()` if you want to change parameters, so you might do this instead with that function

```{r}
dt = 0.25
n_particles <- 10L
sir_model_2 <- gen_sir$new(data = list(dt = 0.25, S_ini = 2000, I_ini = 10, beta = 0.4, gamma = 0.1),
                           step = 1,
                           n_particles = n_particles,
                           n_threads = 4L,
                           seed = 1L)
sir_model_2$state()
```

Let's run this epidemic forward, and plot the trajectories:

```{r fig.height=5, fig.width=7}
n_steps <- 200
x <- array(NA, dim = c(sir_model_2$info()$len, n_particles, n_steps))

for (t in seq_len(n_steps)) {
  x[ , , t] <- sir_model_2$run(t)
}
time <- x[1, 1, ]
x <- x[-1, , ]

par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
matplot(time, t(x[1, , ]), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = "#8c8cd9", lty = 1, ylim = range(x))
matlines(time, t(x[2, , ]), col = "#cc0044", lty = 1)
matlines(time, t(x[3, , ]), col = "#999966", lty = 1)
legend("left", lwd = 1, col = c("#8c8cd9", "#cc0044", "#999966"), legend = c("S", "I", "R"), bty = "n")

> ```

Reviewer: Consider something above like

```
cols <- c(S = "#8c8cd9", I = "#cc0044", R = "#999966")
```

so that you can do

```
matplot(..., col = cols[["S"]])
matlines(..., col = cols[["I"]])
legend(..., col = cols, legend = name(cols))
```

Many other methods are available on the dust object to support inference, but typically you will use the `mcstate` package instead of calling these directly. However, if you wish to simulate state space models with set parameters, this can be done entirely using the commands above from `dust`.

> # Inferring parameters with [`mcstate`](https://mrc-ide.github.io/odin.dust/)

Reviewer: Actually *this* might be where I would start the vignette :)

Author: Yes, starting here

> Typically these models will be used in conjunction with observed data to infer the most likely parameter values, most likely trajectories up to the present (nowcast), and make forecasts using these parameters. The `mcstate` package implements particle filter and pMCMC methods to make all of this this possible from `dust` models.

Reviewer: "most likely" looks incorrect to me as we don't offer any ML methods 😛

Reviewer: Also pMCMC needs defining, possibly link to the wikipedia page or an article?

A comparison function is needed to define a likelihood, i.e. the probability of observing this trajectory, given the data. This is combined with a prior to generate a posterior distribution for each parameter, which in the SIR model gives the distribution of likely values for $\beta$ and $\gamma$. 

> This is straightforward for a deterministic model, but for a stochastic model looking at a single run can easily bias these estimates. `mcstate` therefore implements a particle filter, also known as sequential Monte Carlo (SMC), which runs multiple trajectories, and at each stage where data is available resamples them in proportion to their likelihood. This produces an unbiased estimate of the likelihood for a given set of parameters.

Reviewer: Does a single run "bias" these estimates?

Author: Yes this is wrong, I've read a review briefly and understand this a bit better now

The anatomy of an mcstate particle filter consists of three components:

- A set of observations to fit the model to, generated using `particle_filter_data()`.
- A model to fit, which must be a dust generator.

> - A comparison function, which is an R function which calculates the likelihood of the state given the data.

Reviewer: given the data _at a particular point in time_

## Model data

To demonstrate this, we will attempt to infer the parameters $\beta$ and $\gamma$ using daily case counts. In reality these counts would be measured, but here we generate them from a single run of the model, and adding some constant noise proportional to the epidemic size:

> ```{r}

Reviewer: My suggestion here is that you provide the data as a fait accompli. You can tell the reader that they were generated from the model (or not), but read them from a csv or hide the sausage factory here

Author: Excellent idea, this is confusing

# Set up model
dt <- 0.25
inv_dt <- 1/dt
sir <- gen_sir$new(data = list(dt = dt, S_ini = 1000, I_ini = 10, beta = 0.2, gamma = 0.1),
                   step = 0,
                   n_particles = 1L,
                   n_threads = 1L,
                   seed = 2L)

# Run model forward 100 days, using 4 steps a day
# Calculate the number of new cases by counting how many susceptibles have been depleted between updates
n_steps <- 100
incidence <- rep(NA, length.out = n_steps)
true_history <- array(NA_real_, c(sir$info()$len, 1, n_steps + 1))
true_history[, 1, 1] <- sir$state()
for (t in seq_len(n_steps)) {
    state_start <- sir$state()
    state_end <- sir$run(t * inv_dt)
    true_history[, 1, t + 1] <- state_end
    incidence[t] <- state_start[2, 1] - state_end[2, 1]
}

# Add very simple form of noise
incidence <- incidence + 
  round(rnorm(mean = 0, n = length(true_history[3,1,-1]), sd = 0.1*sqrt(true_history[3,1,-1])))
incidence <- unlist(lapply(incidence, max, 0))
incidence_data <- data.frame(cases = incidence, day = seq_len(n_steps))
incidence
```

This can then be processed with `particle_filter_data()` as follows:

```{r paged.print=TRUE, layout="l-body-outset"}

> library(mcstate)

Reviewer: This is the second inclusion of this library, but I'd prefer ::

                                 time = "day", 
                                 rate = inv_dt)
paged_table(sir_data)
```

More complex models may have additional data streams added as further columns (e.g. deaths, hospital admissions). This will be fed as observed data to the compare function, which will combine and compute the evidence for the current model state.

## Defining the comparison function

The comparison function will typically be a likelihood, the probability of the simulation run given the data. This can be any function in R which takes the current simulated `state`, the state at the previous time step `prev_state`, a `data` row for the current time point and a list of any parameters `pars` the function needs.

For this model the number of cases can be calculated from the change in susceptibles between time steps (as above). In this comparison function, the number of cases within each time interval is modelled as being Poisson distributed. A small amount of noise is added to prevent zero expectations.

```{r}
case_compare <- function(state, prev_state, observed, pars = NULL) {
  if (is.na(observed$cases)) {
    return(NULL)
  }

>   if (is.null(pars$compare$exp_noise)) {

Reviewer: This if / else is complicating things here and was needed I think in order to help with testing. If you want to demonstrate tuning the parameter I would unconditionally use your else branch, otherwise I'd unconditionally use your if branch

matlines(incidence_data$day, t(history[4, , -1]), col = "#cc0044", lty = 1)
matpoints(incidence_data$day, t(true_history[2:4, , -1]), pch = 19,
          col = c("#999966", "#8c8cd9", "#cc0044"))
legend("left", lwd = 1, col = c("#999966", "#8c8cd9", "#cc0044"), 
       legend = c("S", "I", "R"), bty = "n")
```

Doing the same with a $\beta$ and $\gamma$ which are further from the truth yields a lower likelihood:

```{r}
filter$run(save_history = TRUE, pars = list(dt = dt, beta = 0.4, gamma = 0.2))
```

```{r fig.height=5, fig.width=7}
history <- filter$history()
matplot(incidence_data$day, t(history[2, , -1]), type = "l",
        xlab = "Time", ylab = "Number of individuals",
        col = "#999966", lty = 1, ylim = range(history))
matlines(incidence_data$day, t(history[3, , -1]), col = "#8c8cd9", lty = 1)
matlines(incidence_data$day, t(history[4, , -1]), col = "#cc0044", lty = 1)
matpoints(incidence_data$day, t(true_history[2:4, , -1]), pch = 19,
          col = c("#999966", "#8c8cd9", "#cc0044"))
legend("left", lwd = 1, col = c("#999966", "#8c8cd9", "#cc0044"), 
       legend = c("S", "I", "R"), bty = "n")
```

> Typically you will not use the `run()` function of the particle filter directly, and instead use the functions below to infer parameters.

Reviewer: If the user was developing a new inference method, doing a sensitivity analysis, or manually doing a grid search they might?

### Using MCMC to infer parameters

Using this, we can do a MCMC using the Metropolis-Hastings algorithm to infer the values of $\beta$ and $\gamma$ from daily case counts. Let's first describe the parameters we wish to infer, giving a minimum, maximum. For $\gamma$ we will also apply a gamma prior with shape 1 and rate 0.2: $\Gamma(1, 0.2)$:

```{r}
beta <- pmcmc_parameter("beta", 0.2, min = 0)
gamma <- pmcmc_parameter("gamma", 0.1, min = 0, 

>                          prior = function(p) {

Reviewer: it is typical to omit the curly braces here

                           dgamma(p, shape = 1, scale = 0.2, log = TRUE)
                         })

proposal_matrix <- matrix(c(0.01^2, 0, 0, 0.01^2), nrow = 2, ncol = 2, byrow = TRUE)
mcmc_pars <- pmcmc_parameters$new(list(beta = beta, gamma = gamma), proposal_matrix)
```

The proposals for each MCMC step are also set here. Proposals are drawn from a multi-variate normal distribution, with the variance-covariance matrix set here. In this case we have $\beta_{n+1} \sim N(\beta_{n}, 0.01^2)$ and $\gamma_{n+1} \sim N(\gamma_{n}, 0.01^2)$ i.e. no covariance.

Finally, the sampler can be run using these parameters, and the particle filter defined above. Taking 2000 steps with two independent chains:

```{r pmcmc_combined, cache=TRUE}
pmcmc_run <- 
  pmcmc(
    mcmc_pars,
    filter,
    2000,
    save_state = TRUE,
    save_trajectories = TRUE,
    progress = TRUE,
    n_chains = 2
  )
```

We also provide some basic tools for dealing with this output, before running prediction (see below). We can remove burn-in and thin the four chains as follows:

```{r}
processed_chains <- pmcmc_thin(pmcmc_run, burnin = 500, thin = 2)
processed_chains
apply(processed_chains$pars, 2, mean)

> ```

Reviewer: Can we plot something here? It would be nice to see a basic plot of the parameters, and a prediction

Author: I've put in a plot function as above, but the density of trajectories may leave something to be desired. `coda` is used to plot mcmc results (as we removed all plotting functions), and prediction is dealt with in a section below

### Optimal use of parallelisation

> Note that we run four separate chains in serial here. If you have $m$ nodes available, each with $n$ cores, the most efficient way to run this step is to set `nthreads = n` on the filters, and run $m$ independent chains, one on each node setting `n_chains = 1`. These independent results can then be collected using `pmcmc_combine()`:

Reviewer: I'm confused here about introducing both the within-chain parallelism and the multiple chains together

> ```{r pmcmc_serial, cache=TRUE}

Reviewer: The cache will only impact your local run, and even then with horribly unpredictable results. I'll rig something up to make this nicer once it settles down as CRAN will not allow a 30 min vignette. However, perhaps just not running this would be best. TBH I am a bit confused here

pmcmc1 <- 
  pmcmc(
    mcmc_pars,
    filter,
    2000,
    save_state = TRUE,
    save_trajectories = FALSE,
    progress = TRUE,
    n_chains = 1
  )
pmcmc2 <- 
  pmcmc(
    mcmc_pars,
    filter,
    2000,
    save_state = TRUE,
    save_trajectories = FALSE,
    progress = TRUE,
    n_chains = 1
  )
pmcmc3 <- 
  pmcmc(
    mcmc_pars,
    filter,
    2000,
    save_state = TRUE,
    save_trajectories = FALSE,
    progress = TRUE,
    n_chains = 1
  )

> pmcmc_distributed_run <- pmcmc_combine(pmcmc1, pmcmc2, pmcmc3)

Reviewer: Why is this called "distributed" if it was run in series?

Reviewer: ah, ok the idea is that these might have been run across different computers? This feels too much a detail of our workflow for an introductory vignette and might fit better in a vignette on scaling. It also avoids the very real problem of correctly getting the RNG streams correct here.

Reviewer: With all the above, I'd suggest dropping this section for now

Author: Moved to another vignette, we can decide whether to keep or remove separately. I think it's an important advantage, so worth having somewhere from the get go. Agree it's too involved for the intro though

```{r coda_plots, fig.height=10, fig.width=14}
library(coda)
mcmc1 <- as.mcmc(cbind(pmcmc1$probabilities, pmcmc1$pars))
mcmc2 <- as.mcmc(cbind(pmcmc2$probabilities, pmcmc2$pars))
mcmc3 <- as.mcmc(cbind(pmcmc3$probabilities, pmcmc3$pars))
mcmc_sir <- mcmc.list(mcmc1, mcmc2, mcmc3)

# The Rhat statistic may not always be possible to compute
# gelman.diag(mcmc_sir)
rejectionRate(mcmc_sir)
effectiveSize(mcmc_sir)

# Remove burn in and thin
mcmc_sampled <- lapply(list(pmcmc1, pmcmc2, pmcmc3), pmcmc_thin, burnin = 1000, thin = 2)
mcmc1 <- as.mcmc(cbind(mcmc_sampled[[1]]$probabilities, mcmc_sampled[[1]]$pars))
mcmc2 <- as.mcmc(cbind(mcmc_sampled[[2]]$probabilities, mcmc_sampled[[2]]$pars))
mcmc3 <- as.mcmc(cbind(mcmc_sampled[[3]]$probabilities, mcmc_sampled[[3]]$pars))
mcmc_sir <- mcmc.list(mcmc1, mcmc2, mcmc3)

summary(mcmc_sir)
plot(mcmc_sir)
```

## Fitting to multiple datastreams

The set up of the particle filter and observation function makes it easy to fit to multiple datastreams within a Bayesian framework. Let us model a disease where a proportion of infected cases $p_{\mathrm{death}} = 0.05$ die rather than recover by adding the following lines to the odin code:

> ```r

Reviewer: again, dump out the real code here

update(S) <- S - n_SI
update(I) <- I + n_SI - n_IR - n_ID
update(R) <- R + n_IR
update(D) <- D + n_ID

n_IR <- rbinom(I, p_IR * dt * (1 - p_death))
n_ID <- rbinom(I, p_IR * dt * p_death)
n_SI <- rbinom(S, p_SI * dt)

initial(D) <- 0
p_death <- user(0.05)
```

```{r}
gen_death <- odin_dust("deaths.txt")
```

We can simulate some data to fit to, as before. We will also make the case data patchy, making it missing on some days:

```{r layout="l-body-outset", paged.print=TRUE }

> # Set up model

Reviewer: again, use some precooked csv?

inv_dt <- 1/dt
sir_death <- gen_death$new(data = list(dt = dt, S_ini = 1000, I_ini = 10, beta = 0.2, gamma = 0.1),
                   step = 0,
                   n_particles = 1L,
                   n_threads = 1L,
                   seed = 2L)

# Run model forward 100 days, using 4 steps a day
# Calculate the number of new cases by counting how many susceptibles have been depleted between updates
n_steps <- 100
incidence <- rep(NA, length.out = n_steps)
deaths <- rep(NA, length.out = n_steps)
true_history_deaths <- array(NA_real_, c(sir_death$info()$len, 1, n_steps + 1))
true_history_deaths[, 1, 1] <- sir_death$state()
for (t in seq_len(n_steps)) {
    state_start <- sir_death$state()
    state_end <- sir_death$run(t * inv_dt)
    true_history_deaths[, 1, t + 1] <- state_end
    incidence[t] <- state_start[2, 1] - state_end[2, 1]
    deaths[t] <- state_end[5, 1] - state_start[5, 1]
}

# Add very simple form of noise
missing <- rbinom(n = n_steps, prob = 0.05, size = 1) == 1
incidence[missing] <- NA
combined_data <- particle_filter_data(
  data.frame(cases = incidence, deaths = deaths, day = seq_len(n_steps)),
  time = "day",
  rate = inv_dt
)
paged_table(combined_data)
```

> We then need to add this into our comparison function, and in doing so we will also demonstrate the use of the `index` parameter for the particle filter. The only model outputs needed from the comparison function are $S$ and $D$. For models with a very large state space returning only those compartments required by the comparison function can be important for efficiency. An index function can be defined, which operates on dust generators `info()`:

Reviewer: This paragraph hides the motivation in the middle

```{r}
index <- function(info) {
  list(run = c(info$index$S, info$index$D), state = c(info$index$S, info$index$I, info$index$R, info$index$D))
}

index(sir_death$info())
```

Which sends the comparison function the indices listed under `run`, in this case the number of susceptibles, and saves any compartments listed under `state` if the history is kept. In the `compare()` function the index of `state` now corresponds to `index$run`. The missing data must be checked for, and contribute 0 to the likelihood if found:

```{r}
combined_compare <- function(state, prev_state, observed, pars = NULL) {
  if (is.na(observed$cases) && is.na(observed$deaths)) {
    return(NULL)
  }

>   if (is.null(pars$compare$exp_noise)) {

Reviewer: Simplify as above

    exp_noise <- 1e6
  } else {
    exp_noise <- pars$compare$exp_noise
  }
  

>   ll_pois <- function(obs, model, exp_noise) {

Reviewer: This function might be lifted out to be a free function for ease of explanation - looks like it does not need the captured scope?

>     if (is.na(obs)) {

Reviewer: The is.na branch needs explanation

      ll_obs <- numeric(length(model))
    } else {
      lambda <- model +
        rexp(n = length(model), rate = exp_noise)
      ll_obs <- dpois(x = obs, lambda = lambda, log = TRUE)
    }
    ll_obs
  }
  
  ll_cases <- ll_pois(observed$cases, 
                      prev_state[1, , drop = TRUE] - state[1, , drop = TRUE],
                      exp_noise)
  ll_deaths <- ll_pois(observed$deaths, 
                      state[2, , drop = TRUE] - prev_state[2, , drop = TRUE],
                      exp_noise)
  
  ll_cases + ll_deaths
}
```

This can be used to set up a particle filter as before:

```{r}
n_particles <- 100
filter <- particle_filter$new(data = combined_data,
                              model = gen_death,
                              n_particles = n_particles,
                              compare = combined_compare,
                              index = index,
                              n_threads = 4L,
                              seed = 1L)
filter$run(save_history = FALSE, pars = list(dt = dt))
```

and this can be used to run the `pmcmc()` method, as above.

> ## Running predictions

Reviewer: Save the reader bouncing around and put it with the simple model above?

Going back to the simple SIR model above, we can easily continue running the fitted model forward in time using the posterior estimated by `pmcmc()` to create a forecast past the end of the data:

```{r fig.height=5, fig.width=7}
forecast <- predict(processed_chains, 
                    steps = seq(400, 800, 4), 
                    n_threads = 4L, 
                    prepend_trajectories = TRUE,
                    seed = processed_chains$predict$seed)

# Sample 10 particles

> mini_forecast <- forecast$state[,sample.int(dim(forecast$state)[2], size = 10),]

Reviewer: dim(x)[2] is nrow(x)

> time <- mini_forecast[1,1,-1]

Reviewer: Looks like linting is not on here by default. Can you look over the style and fix the bulk of these please? I can do this if you'd rather not

Author: @richfitz linting the vignettes is a mystery to me unfortunately. If I run lintr::lint_package() they are ignored, same with lint_dir. You can run lint on the Rmd files directly and it works, but I don't know how to do this just from the .lintr file.

https://github.com/jimhester/lintr/blob/master/NEWS.md suggests this should run by default as of v2.0.1, so I'm not sure what's going on

Author: Also I am unable to/incorrectly turning off linting of the odin code in CodeFactor

> matplot(time*dt, t(mini_forecast[2, , -1]), type = "l",

Reviewer: same colour treatment as above (and spaces around `*`)
